Firestore/core/src/firebase/firestore/nanopb/message.h

```diff
@@ -130,63 +147,82 @@ class Message {
   }
 
   /**
-   * Serializes this message into a byte buffer.
+   * Returns a pointer to the Nanopb-generated array that describes the fields
+   * of the Nanopb proto; the array is required to call most Nanopb functions.
    *
-   * The lifetime of the return value is entirely independent of this message.
+   * Note that this is essentially a property of the type, but cannot be made
+   * a template parameter for various technical reasons.
    */
-  grpc::ByteBuffer ToByteBuffer() const;
-
- private:
-  // For access to the explicit constructor. Most code shouldn't be able to
-  // modify the underlying proto.
-  friend class remote::WatchStreamSerializer;
-  friend class remote::WriteStreamSerializer;
-  friend class remote::DatastoreSerializer;
-
-  explicit Message(const pb_field_t* fields) : fields_{fields} {
+  static const pb_field_t* fields() {
+    return FieldsArray<T>();
   }
 
-  bool owns_proto() const {
-    return fields_ != nullptr;
+ private:
+  // Important: this function does *not* modify `owns_proto_`.
+  void Free() {
+    if (owns_proto_) {
+      FreeNanopbMessage(fields(), &proto_);
+    }
   }
 
-  // Note: `fields_` doubles as the flag that indicates whether this instance
-  // owns the underlying proto (and consequently should release it upon
-  // destruction).
-  const pb_field_t* fields_ = nullptr;
+  bool owns_proto_ = true;
+  // The Nanopb-proto is value-initialized (zeroed out) to make sure that any
+  // member variables that aren't written to are in a valid state.
   T proto_{};
 };
 
 namespace internal {
+
 util::StatusOr<nanopb::ByteString> ToByteString(const grpc::ByteBuffer& buffer);
+
 }  // namespace internal
 
 template <typename T>
-MaybeMessage<T> Message<T>::TryParse(const pb_field_t* fields,
-                                     const grpc::ByteBuffer& byte_buffer) {
-  auto maybe_bytes = internal::ToByteString(byte_buffer);
-  if (!maybe_bytes.ok()) {
-    return maybe_bytes.status();
-  }
-
-  Message message{fields};
-  nanopb::Reader reader{maybe_bytes.ValueOrDie()};
-  reader.ReadNanopbMessage(fields, message.get());
-  if (!reader.ok()) {
-    return reader.status();
+template <typename U>
+Message<T> Message<T>::TryParse(U* reader) {
+  Message<T> result;
+  reader->ReadNanopbMessage(result.fields(), result.get());
+  if (!reader->ok()) {
+    // Guarantee that a partially-filled message is never returned.
+    return Message<T>{};
   }
+  return result;
+}
 
-  return MaybeMessage<T>{std::move(message)};
+/**
+ * Serializes the given `message` into a `grpc::ByteBuffer`.
+ *
+ * The lifetime of the return value is entirely independent of the `message`.
+ */
+template <typename T>
+grpc::ByteBuffer ToByteBuffer(const Message<T>& message) {
```

Comment:
Would it make more sense as a template?
```cpp
template <typename T, typename U>
T ConvertTo(const Message<U>& message) {
  Writer<T> writer;
  writer.WriteNanopbMessage(message.fields(), message.get());
  return writer.Release();
}

auto grpc_bytes = ConvertTo<grpc::ByteBuffer>(message);
auto nanopb_bytes = ConvertTo<ByteString>(message);
auto str = ConvertTo<std::string>(message);
auto bad = ConvertTo<Foo>(message); // Fails -- no Writer<Foo> defined
```
This would require templatizing `Writer`.

Reply:
Elsewhere for conversion style functions we've just used "Make" + return type (e.g. `MakeNSString`, `MakeByteBuffer`). Overload resolution on the argument type disambiguates conversion from different sources.

Making this a template seems sort of overkill-ish. Can you think of any code that would want to be generic on the kind of representation we want for the rendered message? If not simple classes are better.